lp_solve_4.0/lp_solve.1
.\" 						-*-nroff-*-
.TH LP_SOLVE 1 "November 30, 2002"
.SH NAME
lp_solve \- solve (mixed integer) linear programming problems
.SH SYNOPSIS
\fBlp_solve\fP [\fIoptions\fP] < \fIinput-file\fP
.SH OPTIONS
.TP 1.2i
-v[\fIlevel\fP]
Set verbosity:
.nf
   -v0: CRITICALSTOP
   -v1: CRITICAL
   -v2: SEVERE
   -v3: IMPORTANT (default)
   -v4: NORMAL
   -v5: DETAILED
   -v6: FULL
.fi
If level not provided (-v) then -v4 (NORMAL) is assumed.
-time
Print CPU time to parse input and to calculate result.
.TP
-S\fIdetail\fP
Print solution. If \fIdetail\fP omitted, then -S2 is used.
.nf
   -S0: Print nothing
   -S1: Only objective value
   -S2: Objective value + variables (default)
   -S3: Objective value + variables + constraints
   -S4: Objective value + variables + constraints + duals
   -S5: Objective value + variables + constraints + duals + lp model
   -S6: Objective value + variables + constraints + duals + lp model + lp scales
.fi
.TP
-h
Help mode, prints the usage.
.TP
-d
Debug mode, all intermediate results are printed, and the branch-and-bound
decisions in case of (mixed) integer problems.
.TP
-min
Minimize the objective function. This is the default for MPS input.
In lp_solve format you can specify minimization or maximization in the input
file as well. The command line option overrides.
.TP
-max
Maximize the objective function. This is the default for lp_solve format
input.
In lp_solve format you can specify minimization or maximization in the input
file as well. The command line option overrides.
.TP
-p
Only functional for pure LP problems. Print the values of the dual
variables as well in the result. They are named r_1 until r_XXXXX unless
specified by the user.  Note that bounds (constraints on just one variable)
are not considered real constraints, and are not given a row in the matrix,
and are therefore not printed here.
.TP
-b \fIbound\fP
Specify an upper (when minimizing) or lower (when maximizing) limit for the
value of the objective function to
the program. Only useful for (mixed) integer problems.  If close enough, may
speed up the calculations. The same result can be obtained by adding an extra
constraint to the problem.
.TP
-c
When branching in MILP problems, take the ceiling of the selected non-integer
variable first instead of the floor. This can influence the speed of MILP
problems.
.TP
-B
-B \fIrule\fP
Specify branch-and-bound rule:
.nf
   -B0: Select Lowest indexed non-integer column (default)
   -B1: Select Random non-integer column
   -B2: Select Largest deviation from an integer value
   -B3: Select Best ???
   -B4: Select Median value deviation from an integer value
   -B5: Select Greedy ???
.fi
.TP
-e \fIvalue\fP
\fIvalue\fP is the tolerance of the test for whether the value of a variable
is really integer. \fIvalue\fP must be between 0 and 0.5. Default value is 1e-6
and should be OK for most applications. Of course only useful for MILP
problems.
.TP
-i
Print all intermediate valid solutions. Can give you useful
solutions even if the total run time is too long.
Only useful for (mixed) integer problems.
.TP
-s \fImode\fP
Use automatic problem scaling:
.nf
	  -s:
	 -s0: Numerical range-based scaling
	 -s1: Geometric scaling
	 -s2: Curtis-reid scaling
.fi
This might improve the numerical
stability of your problem.
.TP
-sp
Also do power scaling.
This option must come AFTER -s \fImode\fP option.
.TP
-sl
Also do Lagrange scaling.
This option must come AFTER -s \fImode\fP option.
.TP
-si
Also do Integer scaling.
This option must come AFTER -s \fImode\fP option.
.TP
-I
Print info after reinverting.
.TP
-t
Trace pivot selection.
.TP
-lp
Read from LP file (default).
.TP
-mps
Read from MPS file instead of lp file. For a short introduction to MPS see
ftp://softlib.cs.rice.edu/pub/miplib/mps_format.
.TP
-parse_only
Parse input file but do not calculate (ie check).
.TP
-presolve
Presolve problem before optimizing.
.TP
-improve\fIlevel\fP
Iterative improvement level:
.nf
	 -improve0: none (default)
	 -improve1: FTRAN only
	 -improve2: BTRAN only
	 -improve3: FTRAN + BTRAN
.fi
.TP
-degen
Use random perturbations to reduce degeneracy, can increase numerical
instability.
.TP
-trej \fITrej\fP
Set minimum pivot value to \fITrej\fP.
.SH DESCRIPTION
The linear programming problem can be formulated as: Solve A.x >= V1, with
V2.x maximal. A is a matrix, x a vector of (nonnegative) variables, V1 a
vector called the right hand side, and V2 a vector specifying the objective
function.
.br
Any number of the variables may be specified to be of type integer.
.br
This program solves problems of this kind. It is slightly more general than
the above problem, in that every row of A (specifying one constraint) can have
its own (in)equality, <=, >= or =. The result specifies values for all
variables.
.br
Uses a 'Simplex' algorithm and sparse matrix methods, for pure LP problems.
If one or more of the variables is declared integer, the Simplex algorithm is
iterated with a branch and bound algorithm, until the desired optimal
solution is found.
.br
The "-i" option will print all intermediate valid solutions.
.SH "INPUT SYNTAX"
The default input syntax is a set of algebraic expressions and "int"
declarations in the following order:
.sp
<objective function>
.br
<constraint>+
.br
<declaration>*
.sp
where:
.TP 0.2i
-
<objective function> is a linear combination of variables, ending with a
semicolon, optionally preceded by "max: " or "min: " to indicate whether you
want it to be minimized or maximized. The case is not important, "Max:" or
"MAX:" will work as well. Maximization is the default.
.TP
-
<constraint> is an optional constraint name followed by a colon plus a
linear combination of variables and constants, followed by a relational
operator, followed again by a linear combination of variables and constants,
ending with a semicolon. The relational operator can be any of the following:
"<" "<=" "=" ">" ">=". There is no semantic difference between "<" and "<="
nor between ">" and ">=" (even for integer variables!).
.TP
-
<declaration> is of the form: "int" \fIvar\fP+ ";" Commas are allowed between
variables.
.sp
So, the simplest linear problem consists of an objective function and 1
constraint.
.SH EXAMPLE
The simple problem:
.sp
x1 >= 1
.br
x2 >= 1
.br
x1 + x2 >= 2
.br
minimize x1 + x2 (= maximize -(x1 + x2)), with x1 integer
.sp
can be written as follows:
.sp
-x1 + -x2;
.br
(or min: x1 + x2;)
.br
x1 > 1;
.br
x2 > 1;
.br
x1 + x2 > 2;
.br
int x1;
.sp
The correct result for (x1, x2) is of course (1, 1).
.br
With the -mps option, \fBlp_solve\fP will accept MPS as input format.
.SH BUGS
Specifying a constraint name for a bound (constraints on just single
variables) does not have an effect: they are not stored inside the main matrix
and are not assigned a dual variable.
.TP
-
The problem consists entirely of constraints on just single variables
(so-called "bounds", like x < 1; ) and no constraint with more than 1
variable (like x + 3 y > 17; ). This leaves \fBlp_solve\fP with an empty problem
matrix, as bounds are not stored in the main matrix. No real-life examples
should be of this form, so I am not really chasing this problem.
.TP
-
Many people forget that \fBlp_solve\fP can only handle POSITIVE values for the
variables. While reading MPS files it will however handle free or negative
variables by replacing them with a variable pair \fIvar\fP_neg and \fIvar\fP_pos or
-\fIvar\fP respectively. It is up to the user to interpret the result of this
transformation.
.TP
- Sometimes problems are numerically unstable, and the unavoidable rounding
errors inside \fBlp_solve\fP will cause aborts. It is very hard to give general
solutions to this problem, but try to keep all values in your problem in the
order of magnitude of 1 by proper scaling. This is almost always better than
using \fBlp_solve\fPs built-in scaling (with -s). Almost parallel constraints are
also not very good for numerical stability. Use "lp_solve -v" and observe the
values of the pivots to see if there are any dangerously large or low numbers
there.
.br
Building \fBlp_solve\fP with long doubles (see the Makefile) can help to increase
numerical stability, but will also increase the run time considerably.
.br
You can consult the author as well if you encounter numerical problems, but
please remember that it is very easy to formulate an infeasible LP problem, so
be sure there is a solution.
.SH SEE ALSO
The implementation of the simplex kernel was mainly based on:
.br
W. Orchard-Hays: "Advanced Linear Programming Computing Techniques",
McGraw-Hill 1968
.br
The mixed integer branch and bound part was inspired by:
.br
section 6.4 of "An Introduction to Linear Programming and Game Theory" by
Paul R. Thie, second edition published by John Wiley and Sons in 1988.
.br
This book refers to:
.br
Dakin, R.J., "A Tree Search Algorithm for MILP Problems", Comput. J., 8 (1965)
pp. 250-255
.SH ACKNOWLEDGEMENTS
The work of Jeroen Dirks made the transition from the basic version 1.5 to
the full version 2.0 possible. He contributed the procedural interface, a
built-in MPS reader, and many fixes and enhancements to the code.
.SH CONTRIBUTED BY
M.R.C.M. Berkelaar
.br
Eindhoven University of Technology
.br
Design Automation Section
.br
P.O. Box 513
.br
NL-5600 MB Eindhoven, The Netherlands
.br
phone +31-40-2474792
.br
E-mail: michel@es.ele.tue.nl
.SH STATUS
Use at own risk. Bug reports are welcome, as well as success stories.


lp_solve_4.0/CHANGELOG
Changed after first review for comp.sources.reviewed:

- Bug with row singletons and assignment fixed.
- Makefile fixed.
- Formatted man page (lp_solve.man) added + target to generate it in the
  Makefile
- install signal handler after reading the input to avoid dumping core
- changed names of some examples to make them uniform
- added correct results for all examples

Changes since the comp.sources.reviewed release:

- added more input sanity checking
- removed a bug in the Branch-and-Bound code which could cause umlimited
  recursion and finally a core dump
- removed the duplicate declaration of global variable Rows
- replaced bcopy by the ANSI standard memcopy
- various small cosmetic changes

Changes for version 1.1

- added KNOWN_BUGS file.
- accept x <= 0 as a sane bound.

Changes for version 1.2

- added (unsupported by me) VMS_SUPPORT file
- made lp_solve print the reason a problem cannot be solved (unbounded or
  infeasible)
- added input format extension to optionally indicate whether the objective
  function should be minimized or maximized.
- Added output(!) in real MPS format for the -v option.

Changes for version 1.3

- Added #ifdefs to handle the different definitions necessary for yytext on
  different computers. Why is there no standard???
- Fixed a very nasty bug in the parser which would handle variables on the
  right hand side of the relational operator which had an effective negative
  sign wrongly (actually, the negative sign was ignored).
- added -p option to print the values of the dual variables.
- Added test target in Makefile to allow automatic testing with the standard
  examples ex[1-5].lp.
- Added syntax extension for optional user specified constraint (dual) names.
- Fixed bug in the combination of minimization of the objective function
  (min: <obj fun>) and MILP problems. Pure LP worked ok.
- From now on lp_solve refuses problems with just bounds (0 real constraints).
  This does not work anyway.

Changes for version 1.4

- Various cosmetic changes to the sources
- Added ex6.lp just for fun (read it!).
- Added results obtained on netlib and miplib benchmark sets (thanks to
  mps2eq).
- removed pragmas for Alliant computers (they're out of business anyway).
- changed the default "double" type to REAL, which can be changed during
  compilation (-DREAL=float or -DREAL='long double')

Changes for version 1.5

- Added -scale option for global scaling of the problem.
- Fixed problem with Endetacol being written out of bounds.
- Added -autoscale option for automatic scaling of rows and columns.
- Added ex7.lp, an example which can only be solved when scaled properly.
- Added -solve_dual option to let lp_solve solve the dual of the problem it is
  reading. This is not a very useful option, but I needed it to check the
  results generated by the updated -p option (see next item). It will not work
  with equality constraints.
- The -p option now produces real dual values, thanks to a code hint by
  Jeroen Dirks (jeroend@tor.numetrix.com)
- The MILP branch-and-bound loop was rearranged by Jeroen Dirks resulting in
  significantly increased performance (2 - 10 times faster).


Changes for 2.0a

- Added a procedural interface (the lp toolkit).
- removed -scale and -solve_dual options
- renamed option -autoscale to -s

Changes for 2.0b1

- Added -mps option to read MPS files. And also included some mps associated
  functions in the toolkit.

Changes for 2.0b2

- Fixed bug in:
	delete_lp
	del_constraint
	write_MPS
- added:
	write_LP
	msp2lp
	lp2mps

Changes for 2.0b3

- Changed default integer/noninteger value to 1e-3
- Added Epsb in Iteration function at minit detection
  (lp_solve gave infeasible for some feasible problems)
- Added -degen perturbation option for degenerate problems.

Changes for 2.0

- Removed "Parsing" message
- Made CHECK Warning output print to stderr

Changes for 2.1beta

- added '-time' option to print CPU times of parsing and optimization
- fixed some MPS reading problems
- allow empty OF in lp_solve format to get just a feasible solution
- added filtering of very small pivots to improve numerical stability
- fixed several bugs in lpkit.c and readmps.c

Changes for 2.1

- improved the speed of the MPS reader (considerably!) by introducing hash
  tables
- fixed bug in column_in_lp
- updated Jeroen Dirks' email address in the sources
- added HARTMUT_DOCUMENTATION to the distribution

Changes for 2.2

- fixed memory leak in delete_lp
- fixed some more lpkit.c bugs
- got rid of the ugly global variables used in solve.c.

Changes for 2.3

- fixed memory problem in solve.c (eta arrays were not always realloc-ed on
  time).
- support for RANGES in MPS reworked; should now be according to standard
- option -degen could sometimes produce "results" for infeasible problems,
  fixed.
- previous versions of lp_solve would not scale any column when integer
  variables were present. Now only the actual integer columns are not scaled.
- copy_lp now also copies the rowname and colname hash tables

Changes for 3.0

- lp_solve now has an official open source license, 3.0 and up will have the
  LGPL license attached.
- there was a problem in write_MPS, where numbers were printed with up to 16
  digits instead of the maximum of 12

Changes for 3.0

- Bug fix with -degen option (constants were not so constant any more)

Changes for 3.1

- copy name strings in copy_hash_table to avoid crashes when old lp's are freed
- added call to btran at line 925 of solve.c according to an analysis by Joerg
  Herbers

Changes for 3.1a

- fixed memory management problem in read.c introduced in 3.1

Changes for 3.2

- fixed eta realloc problem in solve.c

Changes for 4.0.0.0

- Added files fortify.c, fortify.h, ufortify.h, declare.h
  fortify is a memory management wrapper. It redirects all memory management
  routines like malloc, calloc, realloc, free, ... to own routines. The purpose
  of this is purely for debugging. This 'library' adds leading and trailing

lp_solve_4.0/lp_solve.c
#include <string.h>
#include <time.h>
#include "lpkit.h"
#include "lpglob.h"
#include "patchlevel.h"

#if defined FORTIFY

int EndOfPgr() {
    exit(1);
    return (0);
}
#endif

void print_help(char *argv[]) {
    printf("Usage of %s version %d.%d.%d.%d:\n", argv[0], MAJORVERSION, MINORVERSION, RELEASE, BUILD);
    printf("%s [options] [[<]input_file]\n", argv[0]);
    printf("List of options:\n");
    printf("-h\t\tprints this message\n");
    printf("-v <level>\tverbose mode, gives flow through the program.\n");
    printf("\t\t if level not provided (-v) then -v2 (NORMAL) is taken.\n");
    printf("\t -v0: CRITICALSTOP\n");
    printf("\t -v1: CRITICAL\n");
    printf("\t -v2: SEVERE\n");
    printf("\t -v3: IMPORTANT (default)\n");
    printf("\t -v4: NORMAL\n");
    printf("\t -v5: DETAILED\n");
    printf("\t -v6: FULL\n");
    printf("-d\t\tdebug mode, all intermediate results are printed,\n\t\tand the branch-and-bound decisions\n");
    printf("-p\t\tprint the values of the dual variables\n");
    printf("-b <bound>\tspecify a lower bound for the objective function\n\t\tto the program. If close enough, may speed up the\n\t\tcalculations.\n");
    printf("-i\t\tprint all intermediate valid solutions.\n\t\tCan give you useful solutions even if the total run time\n\t\tis too long\n");
    printf("-e <number>\tspecifies the epsilon which is used to determine whether a\n\t\tfloating point number is in fact an integer.\n\t\tShould be < 0.5\n");
    printf("-c\t\tduring branch-and-bound, take the ceiling branch first\n");
    printf("-ca\t\tduring branch-and-bound, the algorithm chooses branch\n");
    printf("-B <rule>\tspecify branch-and-bound rule\n");
    printf("\t -B0: Select Lowest indexed non-integer column (default)\n");
    printf("\t -B1: Select Random non-integer column\n");
    printf("\t -B2: Select Largest deviation from an integer value\n");
    printf("\t -B3: Select Best ???\n");
    printf("\t -B4: Select Median value deviation from an integer value\n");
    printf("\t -B5: Select Greedy ???\n");
    printf("-s <mode>\tuse automatic problem scaling.\n");
    printf("\t  -s:\n");
    printf("\t -s0: Numerical range-based scaling\n");
    printf("\t -s1: Geometric scaling\n");
    printf("\t -s2: Curtis-reid scaling\n");
    printf("-sp\t\talso do power scaling.\n\t\tThis option must come AFTER -s <mode> option\n");
    printf("-sl\t\talso do Lagrange scaling.\n\t\tThis option must come AFTER -s <mode> option\n");
    printf("-si\t\talso do Integer scaling.\n\t\tThis option must come AFTER -s <mode> option\n");
    printf("-I\t\tprint info after reinverting\n");
    printf("-t\t\ttrace pivot selection\n");
    printf("-lp\t\tread from LP file (default)\n");
    printf("-mps\t\tread from MPS file\n");
    printf("-degen\t\tuse perturbations to reduce degeneracy,\n\t\tcan increase numerical instability\n");
    printf("-trej <Trej>\tset minimum pivot value\n");
    printf("-parse_only\tparse input file but do not calculate (ie check)\n");
    printf("-presolve\tpresolve problem before start optimizing\n");
    printf("-improve <level>\titerative improvement level\n");
    printf("\t -improve0: none (default)\n");
    printf("\t -improve1: FTRAN only\n");
    printf("\t -improve2: BTRAN only\n");
    printf("\t -improve3: FTRAN + BTRAN\n");
    printf("-time\t\tPrint CPU time to parse input and to calculate result\n");
    printf("-min\t\tMinimize the lp problem (overrules setting in file)\n");
    printf("-max\t\tMaximize the lp problem (overrules setting in file)\n");
    printf("-S <detail>\tPrint solution. If detail ommited, then -S2 is used.\n");
    printf("\t -S0: Print nothing\n");
    printf("\t -S1: Only objective value\n");
    printf("\t -S2: Objective value + variables (default)\n");
    printf("\t -S3: Objective value + variables + constraints\n");
    printf("\t -S4: Objective value + variables + constraints + duals\n");
    printf("\t -S5: Objective value + variables + constraints + duals + lp model\n");
    printf("\t -S6: Objective value + variables + constraints + duals + lp model + lp scales\n");
}

void print_cpu_times(const char *info) {
    static clock_t last_time = 0;
    clock_t new_time;

    new_time = clock();
    fprintf(stderr, "CPU Time for %s: %gs (%gs total since program start)\n",
            info, (new_time - last_time) / (double) CLOCKS_PER_SEC,
            new_time / (double) CLOCKS_PER_SEC);
    last_time = new_time;
}

int main(int argc, char *argv[]) {
    lprec *lp;
    char *filen;
    int i;
    short verbose = IMPORTANT /* CRITICALSTOP */;
    MYBOOL debug = FALSE;
    MYBOOL print_sol = FALSE;
    MYBOOL PRINT_DUALS = FALSE;
    MYBOOL floor_first = TRUE;
    short scaling = 0;
    short print_at_invert = FALSE;
    MYBOOL tracing = FALSE;
    short mps = FALSE;
    MYBOOL anti_degen = FALSE;
    short print_timing = FALSE;
    short parse_only = FALSE;
    MYBOOL do_presolve = FALSE;
    short objective = 0;
    short PRINT_SOLUTION = 2;
    MYBOOL improve = IMPROVE_NONE;
    MYBOOL bb_rule = FIRST_SELECT;
    MYBOOL scalemode = MMSCALING;
    int result;
    REAL obj_bound = (REAL) DEF_INFINITE;
    REAL epsilon = (REAL) DEF_EPSILON;
    REAL epspivot = (REAL) DEF_EPSPIVOT;
    FILE *fpin = stdin;

    /* read command line arguments */

#if defined FORTIFY
    Fortify_EnterScope();
#endif

    for (i = 1; i < argc; i++) {
        if (strncmp(argv[i], "-v", 2) == 0) {
            if (argv[i][2])
                verbose = (short) atoi(argv[i] + 2);
            else
                verbose = NORMAL;
        } else if (strcmp(argv[i], "-d") == 0)
            debug = TRUE;
        else if (strcmp(argv[i], "-i") == 0)
            print_sol = TRUE;
        else if (strcmp(argv[i], "-c") == 0)
            floor_first = FALSE;
        else if (strcmp(argv[i], "-ca") == 0)
            floor_first = AUTOMATIC;
        else if (strncmp(argv[i], "-B", 2) == 0) {
            if (argv[i][2])
                bb_rule = (MYBOOL) atoi(argv[i] + 2);
            else
                bb_rule = FIRST_SELECT;
        } else if (strcmp(argv[i], "-b") == 0)
            obj_bound = atof(argv[++i]);
        else if (strcmp(argv[i], "-e") == 0) {
            epsilon = atof(argv[++i]);
            if ((epsilon <= 0.0) || (epsilon >= 0.5)) {
                fprintf(stderr, "Invalid epsilon %g; 0 < epsilon < 0.5\n",
                        (double) epsilon);
                exit(EXIT_FAILURE);
            }
        } else if (strcmp(argv[i], "-p") == 0)
            PRINT_DUALS = TRUE;
        else if (strcmp(argv[i], "-h") == 0) {
            print_help(argv);
            exit(EXIT_SUCCESS);
        } else if (strcmp(argv[i], "-sp") == 0)
            scalemode = (MYBOOL) (scalemode | POWERSCALE);
        else if (strcmp(argv[i], "-sl") == 0)
            scalemode = (MYBOOL) (scalemode | LAGRANGESCALE);
        else if (strcmp(argv[i], "-s2") == 0)
            scaling = 2;
        else if (strcmp(argv[i], "-si") == 0)
            scalemode = (MYBOOL) (scalemode | INTEGERSCALE);
        else if (strncmp(argv[i], "-s", 2) == 0) {
            scaling = 1;
            if (argv[i][2])
                scalemode = (MYBOOL) atoi(argv[i] + 2);
            else
                scalemode = MMSCALING;
        } else if (strcmp(argv[i], "-I") == 0)
            print_at_invert = TRUE;
        else if (strcmp(argv[i], "-t") == 0)
            tracing = TRUE;
        else if (strncmp(argv[i], "-S", 2) == 0) {
            if (argv[i][2])
                PRINT_SOLUTION = (short) atoi(argv[i] + 2);
            else
                PRINT_SOLUTION = 2;
        } else if (strncmp(argv[i], "-improve", 8) == 0) {
            if (argv[i][8])
                improve = (MYBOOL) atoi(argv[i] + 8);
            else
                improve = 0;
        } else if (strcmp(argv[i], "-mps") == 0)
            mps = TRUE;
        else if (strcmp(argv[i], "-lp") == 0)
            mps = FALSE;
        else if (strcmp(argv[i], "-degen") == 0)
            anti_degen = TRUE;
        else if (strcmp(argv[i], "-time") == 0) {
            if (clock() == -1)
                fprintf(stderr, "CPU times not available on this machine\n");
            else
                print_timing = TRUE;
        } else if (strcmp(argv[i], "-trej") == 0)
            epspivot = atof(argv[++i]);
        else if (strcmp(argv[i], "-parse_only") == 0)
            /* only useful for parser software development */
            parse_only = TRUE;
        else if (strcmp(argv[i], "-presolve") == 0)
            do_presolve = TRUE;
        else if (strcmp(argv[i], "-min") == 0)
            objective = -1;
        else if (strcmp(argv[i], "-max") == 0)
            objective = 1;
        else if (fpin == stdin) {
            filen = argv[i];
            if (*filen == '<')
                filen++;
            if ((fpin = fopen(filen, "r")) == NULL) {
                fprintf(stderr, "Error, Unable to open input file '%s'\n",
                        argv[i]);
                print_help(argv);
                exit(EXIT_FAILURE);
            }
        } else {
            filen = argv[i];
            if (*filen != '>') {
                fprintf(stderr, "Error, Unrecognized command line argument '%s'\n",
                        argv[i]);
                print_help(argv);
                exit(EXIT_FAILURE);
            }
        }
    }

    if (mps)
        lp = read_mps(fpin, verbose);
    else /* standard lp_solve syntax expected */
        lp = read_lp_file(fpin, verbose, "lp");

    if (fpin != stdin)
        fclose(fpin);

    if (print_timing)
        print_cpu_times("Parsing input");

    if (parse_only)
        exit(0);

    if (lp == NULL) {
        fprintf(stderr, "Unable to read model.\n");
        exit(EXIT_FAILURE);
    }

    if (objective != 0) {
        if (objective == 1)
            set_maxim(lp);
        else
            set_minim(lp);
    }

    if (PRINT_SOLUTION >= 5)
        print_lp(lp);

    set_print_sol(lp, print_sol);
    set_epsilon(lp, epsilon);
    set_epspivot(lp, epspivot);
    set_print_duals(lp, PRINT_DUALS);
    set_debug(lp, debug);
    set_floor_first(lp, floor_first);
    set_print_at_invert(lp, print_at_invert);
    set_trace(lp, tracing);
    if (obj_bound != DEF_INFINITE)
        set_obj_bound(lp, obj_bound);
    set_anti_degen(lp, anti_degen);
    set_do_presolve(lp, do_presolve);
    set_improve(lp, improve);
    set_scalemode(lp, scalemode);
    set_bb_rule(lp, bb_rule);

    if (scaling == 1)
        auto_scale(lp);
    if (scaling == 2)
        scaleCR(lp);

    if (PRINT_SOLUTION >= 6)
        print_scales(lp);

    result = solve(lp);

    if (print_timing)
        print_cpu_times("solving");

    if (result == OPTIMAL) {
        if (PRINT_SOLUTION >= 1)
            print_objective(lp);

        if (PRINT_SOLUTION >= 2)
            print_solution(lp);

        if (PRINT_SOLUTION >= 3)
            print_constraints(lp);

        if (PRINT_SOLUTION >= 4)
            print_duals(lp);

        if (tracing)
            fprintf(stderr,
                "Branch & Bound depth: %d\nNodes processed: %d\nSimplex pivots: %d\n",
                get_max_level(lp), get_total_nodes(lp), get_total_iter(lp));
    } else if (result == INFEASIBLE) {
        if (PRINT_SOLUTION >= 1)
            printf("This problem is infeasible\n");
    } else if (result == UNBOUNDED) {
        if (PRINT_SOLUTION >= 1)
            printf("This problem is unbounded\n");
    } else if (result == FAILURE) {
        if (PRINT_SOLUTION >= 1)
            printf("lp_solve failed\n");
    }

    delete_lp(lp);

#if defined FORTIFY
    Fortify_LeaveScope();
#endif

    return (result);
}
